Keep creating namespaces

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->


**What happened**:
After this https://github.com/karmada-io/karmada/pull/2825 merged, the namespace will be continued creating.

**What you expected to happen**:
Not continue creating.

**How to reproduce it (as minimally and precisely as possible)**:
I checked the code, and the function `EnsureNamespaceExist` is only invoked when cluster init(karmada-agent init and karmadactl init).

And I tested release-1.4 branch, init clusters with `local-up-karmada.sh` and` karmadactl init`, both works correctly, but the problem is not reproduced.
@yizhang-zen, can you give more information about this? how to reproduce this problem?

**Anything else we need to know?**:
none

**Environment**:
- Karmada version: release-1.4 branch
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):release-1.4 branch
- Others: none

Hi @jwcesign:
The reason why we see this issue is because we have opa-gatekeeper policies that check the compliance of resources (e.g. namespace, clusterrole, clusterrolebinding).
The code is attempting to create a namespace (which our gatekeeper validation webhook catches). Even if we create the namespace prior to the agent init, it will still attempt to create the namespace when apply the deployment below (but it is a normal deployment with no major tweaks).
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: karmada
  namespace: karmada-system
  labels:    
    app: karmada-agent 
    team: compute
    project: sfn_kubernetes
spec:
  replicas: 1
  selector:
    matchLabels:      
      app: karmada-agent
  strategy:
        rollingUpdate:
          maxSurge: 50%
          maxUnavailable: 0
        type: RollingUpdate
  template:
    metadata:
      labels:        
        app: karmada-agent    
        team: compute
        project: sfn_kubernetes    
    spec:
      serviceAccountName: karmada
      containers:
        - name: karmada
          image: docker.io/karmada/karmada-agent:latest
          imagePullPolicy: Always
          command:
            - /bin/karmada-agent
            - --karmada-kubeconfig=/etc/kubeconfig/kubeconfig
            - --cluster-name=hub
            - --cluster-api-endpoint=https://kubernetes.default.svc.cluster.local/
            - --cluster-status-update-frequency=10s
            - --leader-elect-resource-namespace=karmada-system
            - --bind-address=0.0.0.0
            - --secure-port=10357
            - --v=4
          livenessProbe:
            httpGet:
              path: /healthz
              port: 10357
              scheme: HTTP
            failureThreshold: 3
            initialDelaySeconds: 15
            periodSeconds: 15
            timeoutSeconds: 5
          volumeMounts:
            - name: kubeconfig
              mountPath: /etc/kubeconfig
          resources:
            {}
      volumes:
        - name: kubeconfig
          secret:
            secretName: karmada-kubeconfig
```

agent initialization error:
```
I1221 17:13:19.965615       1 agent.go:115] karmada-agent version: version.Info{GitVersion:"v1.4.0-104-g80e5bd9e", GitCommit:"80e5bd9ee9646154fda36250adbcf6719a336290", GitTreeState:"clean", BuildDate:"2022-12-21T06:37:11Z", GoVersion:"go1.19.4", Compiler:"gc", Platform:"linux/amd64"}
 "command failed" err="ensure namespace failed due to create failed. namespace: karmada-cluster, error: admission webhook \"validation.gatekeeper.sh\" denied the request: [globals-must-meet-compliance] zendesk/opa-gatekeeper Namespace  karmada-cluster: you must provide labels {\"compliance\"}. See <REDACTED>."
```
Creating the Namespace upfront (i.e. prior to agent initialization) and using version [docker.io/karmada/karmada-agent:v1.3.3](http://docker.io/karmada/karmada-agent:v1.3.3), it will not attempt to create the namespace and all good

--- a/pkg/util/namespace.go
+++ b/pkg/util/namespace.go
@@ -10,6 +10,20 @@ import (
 	kubeclient "k8s.io/client-go/kubernetes"
 )
 
+// IsNamespaceExist tells if the namespace already exists.
+func IsNamespaceExist(client kubeclient.Interface, namespace string) (bool, error) {
+	_, err := client.CoreV1().Namespaces().Get(context.Background(), namespace, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	return true, nil
+}
+
 // CreateNamespace just try to create the namespace.
 func CreateNamespace(client kubeclient.Interface, namespaceObj *corev1.Namespace) (*corev1.Namespace, error) {
 	_, err := client.CoreV1().Namespaces().Create(context.TODO(), namespaceObj, metav1.CreateOptions{})
@@ -40,6 +54,16 @@ func EnsureNamespaceExist(client kubeclient.Interface, namespace string, dryRun 
 	namespaceObj.Name = namespace
 
 	if dryRun {
+		return namespaceObj, nil
+	}
+
+	// It's necessary to check if a namespace exists before creating it.
+	// Sometimes the namespace is created in advance, to give less privilege to Karmada.
+	exist, err := IsNamespaceExist(client, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if namespace exist. namespace: %s, error: %v", namespace, err)
+	}
+	if exist {
 		return namespaceObj, nil
 	}
 

--- a/pkg/util/namespace_test.go
+++ b/pkg/util/namespace_test.go
@@ -14,6 +14,70 @@ import (
 	coretesting "k8s.io/client-go/testing"
 )
 
+func TestIsNamespaceExist(t *testing.T) {
+	type args struct {
+		client    *fake.Clientset
+		namespace string
+		reactor   coretesting.ReactionFunc
+	}
+
+	tests := []struct {
+		name    string
+		args    args
+		want    bool
+		wantErr bool
+	}{
+		{
+			name: "query namespace error",
+			args: args{
+				client:    fake.NewSimpleClientset(&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}),
+				namespace: metav1.NamespaceDefault,
+				reactor: func(action coretesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &corev1.Namespace{}, errors.New("failed to get namespace")
+				},
+			},
+			want:    false,
+			wantErr: true,
+		},
+		{
+			name: "namespace not exists",
+			args: args{
+				client:    fake.NewSimpleClientset(&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default-1"}}),
+				namespace: metav1.NamespaceDefault,
+				reactor:   nil,
+			},
+			want:    false,
+			wantErr: false,
+		},
+		{
+			name: "namespace already exists",
+			args: args{
+				client:    fake.NewSimpleClientset(&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}),
+				namespace: metav1.NamespaceDefault,
+				reactor:   nil,
+			},
+			want:    true,
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.args.reactor != nil {
+				tt.args.client.PrependReactor("get", "namespaces", tt.args.reactor)
+			}
+
+			got, err := IsNamespaceExist(tt.args.client, tt.args.namespace)
+			if (err == nil && tt.wantErr == true) || (err != nil && tt.wantErr == false) {
+				t.Errorf("IsNamespaceExist() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("IsNamespaceExist() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
 func TestCreateNamespace(t *testing.T) {
 	type args struct {
 		client       *fake.Clientset
@@ -135,6 +199,7 @@ func TestEnsureNamespaceExist(t *testing.T) {
 		client        *fake.Clientset
 		namespace     string
 		dryRun        bool
+		reactorGet    coretesting.ReactionFunc
 		reactorCreate coretesting.ReactionFunc
 	}
 	tests := []struct {
@@ -163,10 +228,21 @@ func TestEnsureNamespaceExist(t *testing.T) {
 			wantErr: false,
 		},
 		{
+			name: "check namespace exists error",
+			args: args{
+				client:    fake.NewSimpleClientset(&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}),
+				namespace: metav1.NamespaceDefault,
+				reactorGet: func(action coretesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, nil, errors.New("failed to get namespace")
+				},
+			},
+			wantErr: true,
+		},
+		{
 			name: "create namespace error",
 			args: args{
 				client:    fake.NewSimpleClientset(),
-				namespace: "default",
+				namespace: metav1.NamespaceDefault,
 				reactorCreate: func(action coretesting.Action) (handled bool, ret runtime.Object, err error) {
 					return true, nil, errors.New("failed to create namespace")
 				},
@@ -176,6 +252,10 @@ func TestEnsureNamespaceExist(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			if tt.args.reactorGet != nil {
+				tt.args.client.PrependReactor("get", "namespaces", tt.args.reactorGet)
+			}
+
 			if tt.args.reactorCreate != nil {
 				tt.args.client.PrependReactor("create", "namespaces", tt.args.reactorCreate)
 			}

--- a/pkg/util/serviceaccount.go
+++ b/pkg/util/serviceaccount.go
@@ -26,6 +26,20 @@ func CreateServiceAccount(client kubeclient.Interface, saObj *corev1.ServiceAcco
 	return saObj, nil
 }
 
+// IsServiceAccountExist tells if specific service account already exists.
+func IsServiceAccountExist(client kubeclient.Interface, namespace string, name string) (bool, error) {
+	_, err := client.CoreV1().ServiceAccounts(namespace).Get(context.Background(), name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	return true, nil
+}
+
 // DeleteServiceAccount just try to delete the ServiceAccount.
 func DeleteServiceAccount(client kubeclient.Interface, namespace, name string) error {
 	err := client.CoreV1().ServiceAccounts(namespace).Delete(context.Background(), name, metav1.DeleteOptions{})
@@ -39,6 +53,14 @@ func DeleteServiceAccount(client kubeclient.Interface, namespace, name string) e
 // If service account not exit, just create it.
 func EnsureServiceAccountExist(client kubeclient.Interface, serviceAccountObj *corev1.ServiceAccount, dryRun bool) (*corev1.ServiceAccount, error) {
 	if dryRun {
+		return serviceAccountObj, nil
+	}
+
+	exist, err := IsServiceAccountExist(client, serviceAccountObj.Namespace, serviceAccountObj.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if service account exist. service account: %s/%s, error: %v", serviceAccountObj.Namespace, serviceAccountObj.Name, err)
+	}
+	if exist {
 		return serviceAccountObj, nil
 	}
 

--- a/pkg/util/serviceaccount_test.go
+++ b/pkg/util/serviceaccount_test.go
@@ -27,7 +27,7 @@ func TestCreateServiceAccount(t *testing.T) {
 		wantErr bool
 	}{
 		{
-			name: "already exist",
+			name: "service account already exist",
 			args: args{
 				client: fake.NewSimpleClientset(makeServiceAccount("test")),
 				sa:     makeServiceAccount("test"),
@@ -36,7 +36,7 @@ func TestCreateServiceAccount(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name: "create error",
+			name: "service account create error",
 			args: args{
 				client: alwaysErrorKubeClient,
 				sa:     makeServiceAccount("test"),
@@ -45,7 +45,7 @@ func TestCreateServiceAccount(t *testing.T) {
 			wantErr: true,
 		},
 		{
-			name: "create success",
+			name: "service account create success",
 			args: args{
 				client: fake.NewSimpleClientset(),
 				sa:     makeServiceAccount("test"),
@@ -80,7 +80,7 @@ func TestDeleteServiceAccount(t *testing.T) {
 		wantErr bool
 	}{
 		{
-			name: "not found",
+			name: "service account not found",
 			args: args{
 				client:    fake.NewSimpleClientset(),
 				namespace: metav1.NamespaceDefault,
@@ -89,7 +89,7 @@ func TestDeleteServiceAccount(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name: "error",
+			name: "service account delete failed",
 			args: args{
 				client:    alwaysErrorKubeClient,
 				namespace: metav1.NamespaceDefault,
@@ -98,7 +98,7 @@ func TestDeleteServiceAccount(t *testing.T) {
 			wantErr: true,
 		},
 		{
-			name: "success",
+			name: "service account delete success",
 			args: args{
 				client:    fake.NewSimpleClientset(makeServiceAccount("test")),
 				namespace: metav1.NamespaceDefault,
@@ -139,7 +139,7 @@ func TestEnsureServiceAccountExist(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name: "exist",
+			name: "service account already exist",
 			args: args{
 				client:            fake.NewSimpleClientset(makeServiceAccount("test")),
 				serviceAccountObj: makeServiceAccount("test"),
@@ -149,7 +149,7 @@ func TestEnsureServiceAccountExist(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name: "not exist",
+			name: "service account not exists",
 			args: args{
 				client:            fake.NewSimpleClientset(),
 				serviceAccountObj: makeServiceAccount("test"),
@@ -159,7 +159,21 @@ func TestEnsureServiceAccountExist(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name: "create error",
+			name: "get service account error",
+			args: args{
+				client: func() kubernetes.Interface {
+					c := fake.NewSimpleClientset()
+					c.PrependReactor("get", "*", errorAction)
+					return c
+				}(),
+				serviceAccountObj: makeServiceAccount("test"),
+				dryRun:            false,
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "create service account error",
 			args: args{
 				client: func() kubernetes.Interface {
 					c := fake.NewSimpleClientset()
@@ -182,6 +196,63 @@ func TestEnsureServiceAccountExist(t *testing.T) {
 			}
 			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("EnsureServiceAccountExist() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestIsServiceAccountExist(t *testing.T) {
+	type args struct {
+		client    kubernetes.Interface
+		namespace string
+		name      string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    bool
+		wantErr bool
+	}{
+		{
+			name: "service account not found",
+			args: args{
+				client:    fake.NewSimpleClientset(),
+				namespace: metav1.NamespaceDefault,
+				name:      "test",
+			},
+			want:    false,
+			wantErr: false,
+		},
+		{
+			name: "kube client return error",
+			args: args{
+				client:    alwaysErrorKubeClient,
+				namespace: metav1.NamespaceDefault,
+				name:      "test",
+			},
+			want:    false,
+			wantErr: true,
+		},
+		{
+			name: "service account already exist",
+			args: args{
+				client:    fake.NewSimpleClientset(makeServiceAccount("test")),
+				namespace: metav1.NamespaceDefault,
+				name:      "test",
+			},
+			want:    true,
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := IsServiceAccountExist(tt.args.client, tt.args.namespace, tt.args.name)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("IsServiceAccountExist() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("IsServiceAccountExist() got = %v, want %v", got, tt.want)
 			}
 		})
 	}

